plugin/manager.go

```diff
@@ -96,3 +216,62 @@ func (m *CLIManager) Uninstall(ctx context.Context, name string) error {
 	}
 	return os.RemoveAll(pluginDirPath)
 }
+
+// parsePluginFromDir checks if a dir is a valid plugin dir which contains
+// one and only one plugin executable file. The dir may contain extra lib files
+// and LICENSE files. Sub-directories are ignored.
+//
+// On success, the plugin executable file path, plugin name and
+// nil error are returned.
+func parsePluginFromDir(path string) (string, string, error) {
+	// sanity check
+	fi, err := os.Stat(path)
+	if err != nil {
+		return "", "", err
+	}
+	if !fi.Mode().IsDir() {
+		return "", "", file.ErrNotDirectory
+	}
+	// walk the path
+	var pluginExecutableFile, pluginName string
+	var foundPluginExecutableFile bool
+	if err := filepath.WalkDir(path, func(p string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		// skip sub-directories
+		if d.IsDir() && d.Name() != filepath.Base(path) {
+			return fs.SkipDir
+		}
+		info, err := d.Info()
+		if err != nil {
+			return err
+		}
+		// only take regular files
+		if info.Mode().IsRegular() {
+			if pluginName, err = ParsePluginName(d.Name()); err != nil {
+				// file name does not follow the notation-{plugin-name} format,
+				// continue
+				return nil
+			}
+			isExec, err := isExecutableFile(p)
+			if err != nil {
+				return err
+			}
+			if isExec {
+				if foundPluginExecutableFile {
+					return errors.New("found more than one plugin executable files")
+				}
+				foundPluginExecutableFile = true
+				pluginExecutableFile = p
+			}
```

Comment:
nit: code style: avoid nesting if
```suggestion
			if !isExec {
				return nil
			}
			if foundPluginExecutableFile {
				return errors.New("found more than one plugin executable files")
			}
			foundPluginExecutableFile = true
			pluginExecutableFile = p

```